macros/contexts/contextNondecimalBase.pl

```diff
@@ -0,0 +1,273 @@
+	#            class => 'context::Hex::UOP::not'},
+	# );
+	# $context->{precedence}{Hex} = $context->{precedence}{special};
+	$context->flags->set(limits => [ -1000, 1000, 1 ]);
+	$context->update;
+
+	# main::PG_restricted_eval('sub Hex {context::Hex::Hex->new(@_)}');
+}
+use Data::Dumper;
+
+sub convert {
+	my $value = shift;
+	print Dumper "convert: $value";
+	# Set default options and get passed in options.
+	my %options = (
+		from   => 10,
+		to     => 10,
+		digits => $digits16,
+		@_
+	);
+	my $from   = $options{'from'};
+	my $to     = $options{'to'};
+	my $digits = $options{'digits'};
+	print Dumper "from: $from; to: $to";
+
+	die "The digits option must be an array of characters to use for the digits"
+		unless ref($digits) eq 'ARRAY';
+
+	#
+	# The highest base the digits will support
+	#
+	my $maxBase = scalar(@$digits16);
+
+	die "The base of conversion must be between 2 and $maxBase"
+		unless $to >= 2 && $to <= $maxBase && $from >= 2 && $from <= $maxBase;
+
+	#
+	# Reverse map the digits to base 10 values
+	#
+	my $baseBdigits = { map { ($digits->[$_], $_) } (0 .. $from - 1) };
+
+	#
+	#  Convert to base 10
+	#
+	my $base10;
+	if ($from == 10) {
+		die "The number to convert must consist only of digits: 0,1,2,3,4,5,6,7,8,9"
+			unless $value =~ m/^\d+$/;
+		$base10 = $value;
+	} else {
+		$base10 = 0;
+		foreach my $d (split(//, $value)) {
+			die "The number to convert must consist only of digits: " . join(',', @$digits[ 0 .. $from - 1 ])
+				unless defined($baseBdigits->{$d});
+			$base10 = $base10 * $from + $baseBdigits->{$d};
+		}
+	}
+	return $base10 if $to == 10;
+
+	#
+	#  Convert to desired base
+	#
+	my @base;
+	do {
+		my $d = $base10 % $to;
+		$base10 = ($base10 - $d) / $to;
+		unshift(@base, $digits->[$d]);
+	} while $base10;
+
+	return join('', @base);
+}
+
+# set the base for the context.
+
+use Data::Dumper;
+
+sub setBase {
+	my $b = shift;
+	return Value::Error('The base must be greater than 1 and less than or equal to $max_base')
+		unless $b >= 2 && $b <= scalar(@$digits16);
+	$base = $b;
+	# $digits16 = [map { $digits16->[$_] } (0..($base-1))];
+	# $digit16 = { map { ($digits16->[$_], $_) } (0 .. scalar(@$digits16) - 1) };
+	# $context->{pattern}{number} = '[' . join('',@$digits16) . ']+';
+}
+
+###########################################################################
+#
+#  A replacement for Parser::Number that acepts numbers in
+#  hexadecimal and converts them to decimal for internal use
+#
+package context::NondecimalBase::Number;
+our @ISA = ('Parser::Number');
+
+use Data::Dumper;
+
+# Create a new number in the given base and convert to base 10.
+
+sub new {
+	my $self = shift;
+	my ($equation, $value, $ref) = @_;
+	$value = context::NondecimalBase::convert($value, from => $base);
+	return $self->SUPER::new($equation, $value, $ref);
+}
+
+#
+#  Return the value of the number in its given base.
+#
+sub eval {
+	$self = shift;
+	return context::NondecimalBase::convert($self->{value}, to => $base);
```

Comment:
This should be removed so that the `eval()` returns the proper `Value::Real` subclass (in the following line), not a string.  Then make sure that `$context->{value}{Real}` is properly set above to `context::NondecimalBase::Real`.